# Chapter 5: Powers, Roots, Exponentials, and Logarithms

**Level:** Applied  
**Prerequisites:** Chapters 1 to 4  
**Estimated study time:** 80 to 110 minutes

Objectives: calculate powers and roots, compare linear and exponential change, interpret logarithms, and solve time-to-target problems.

**Data statement:** All Nusantara Rasa Global examples are fictional and synthetic.

In [1]:
import math
import numpy as np
import matplotlib.pyplot as plt

def compound_value(initial, rate, periods):
    if initial < 0 or rate <= -1:
        raise ValueError("invalid exponential domain")
    return initial * (1 + rate) ** periods

def log_base(value, base):
    if value <= 0 or base <= 0 or base == 1:
        raise ValueError("invalid logarithm domain")
    return math.log(value) / math.log(base)

def time_to_target(initial, target, rate):
    if initial <= 0 or target <= 0 or rate <= -1 or rate == 0:
        raise ValueError("invalid time-to-target domain")
    return math.log(target/initial) / math.log(1+rate)

## Situation 1: Compound demand growth

In [2]:
print(f"Year 2 demand: {compound_value(10000,0.08,2):,.0f} cases")
print(f"Year 5 demand: {compound_value(10000,0.08,5):,.0f} cases")

Year 2 demand: 11,664 cases
Year 5 demand: 14,693 cases


## Situation 2: Linear and exponential scenarios

In [3]:
years=np.arange(0,11)
linear=10000+800*years
exponential=10000*(1.08**years)
print(f"Year 5 linear: {linear[5]:,.0f}")
print(f"Year 5 exponential: {exponential[5]:,.0f}")
fig,ax=plt.subplots(figsize=(7,4))
ax.plot(years,linear,marker="o",label="Add 800 yearly")
ax.plot(years,exponential,marker="o",label="Grow 8% yearly")
ax.set(xlabel="Year",ylabel="Demand (cases)",title="Linear and exponential NRG demand scenarios")
ax.grid(alpha=0.3)
ax.legend()
fig.tight_layout()

Year 5 linear: 14,000
Year 5 exponential: 14,693


The linear scenario repeats an absolute increase. The exponential scenario repeats a multiplicative factor.

## Situation 3: Logarithms ask for exponents

In [4]:
doubling=time_to_target(10000,20000,0.08)
print(f"log10(1000) = {log_base(1000,10):.0f}")
print(f"Continuous doubling time: {doubling:.2f} years")
print(f"First whole year above target: {math.ceil(doubling)}")

log10(1000) = 3
Continuous doubling time: 9.01 years
First whole year above target: 10


The continuous model crosses after about 9.01 years. A year-end planning rule first reports the target exceeded at year 10.

## Situation 4: Domain protection

In [5]:
try:
    log_base(0,10)
except ValueError as error:
    print(error)

invalid logarithm domain


## Practice

Calculate 12,000 growing at 6% for four periods, then plot periods 0 through 8.

In [ ]:
# Write your solution here
